vsintegration/src/FSharp.Editor/FSharpColorizationService.fs

```diff
@@ -0,0 +1,141 @@
+        let mutable runningLexState = ref(0L)
+        let result = new List<ClassifiedSpan>()
+        let sourceTokenizer = FSharpSourceTokenizer(defines, fileName)
+        
+        let compilerTokenToRoslynToken(colorKind: FSharpTokenColorKind) = 
+            match colorKind with
+            | FSharpTokenColorKind.Comment -> ClassificationTypeNames.Comment
+            | FSharpTokenColorKind.Identifier -> ClassificationTypeNames.Identifier
+            | FSharpTokenColorKind.Keyword -> ClassificationTypeNames.Keyword
+            | FSharpTokenColorKind.String -> ClassificationTypeNames.StringLiteral
+            | FSharpTokenColorKind.Text -> ClassificationTypeNames.Text
+            | FSharpTokenColorKind.UpperIdentifier -> ClassificationTypeNames.Identifier
+            | FSharpTokenColorKind.Number -> ClassificationTypeNames.NumericLiteral
+            | FSharpTokenColorKind.InactiveCode -> ClassificationTypeNames.ExcludedCode 
+            | FSharpTokenColorKind.PreprocessorKeyword -> ClassificationTypeNames.PreprocessorKeyword 
+            | FSharpTokenColorKind.Operator -> ClassificationTypeNames.Operator
+            | FSharpTokenColorKind.TypeName  -> ClassificationTypeNames.ClassName
+            | FSharpTokenColorKind.Default | _ -> ClassificationTypeNames.Text
+
+        let scanNextToken(lineTokenizer: FSharpLineTokenizer, colorMap: string[], lexState: Ref<int64>) =
+            let tokenInfoOption, currentLexState = lineTokenizer.ScanToken(lexState.Value)
+            lexState.Value <- currentLexState
+            match tokenInfoOption with
+            | None -> false
+            | Some(tokenInfo) ->
+                let classificationType = compilerTokenToRoslynToken(tokenInfo.ColorClass)
+                for i = tokenInfo.LeftColumn to tokenInfo.RightColumn do
+                    Array.set colorMap i classificationType
+                true
+
+        let scanSourceLine(textLine: TextLine, lexState: Ref<int64>) =
+            let lineTokenizer = sourceTokenizer.CreateLineTokenizer(textLine.Text.ToString(textLine.Span))
+            let colorMap = Array.create textLine.Span.Length ClassificationTypeNames.Text
+            while scanNextToken(lineTokenizer, colorMap, lexState) do ()
+
+            let mutable startPosition = 0
+            let mutable endPosition = startPosition
+            while startPosition < colorMap.Length do
+                let classificationType = colorMap.[startPosition]
+                endPosition <- startPosition
+                while endPosition < colorMap.Length && classificationType = colorMap.[endPosition] do
+                    endPosition <- endPosition + 1
+                let textSpan = new TextSpan(textLine.Start + startPosition, endPosition - startPosition)
+                result.Add(new ClassifiedSpan(classificationType, textSpan))
+                startPosition <- endPosition
+
+        for i = 0 to sourceText.Lines.Count - 1 do
```

Comment:
But what do we know about the initial lex state of that text (which the code above assume to be 0L, which is the initial lex state for the head of a file) ?  What if that text begins in the middle of a ``#if`` or a string or a comment?